tests/src/whisk/core/database/test/CouchDbRestClientTests.scala

```diff
@@ -0,0 +1,140 @@
+
+    val config = new WhiskConfig(Map(
+        dbProvider -> null,
+        dbProtocol -> null,
+        dbUsername -> null,
+        dbPassword -> null,
+        dbHost -> null,
+        dbPort -> null))
+
+    val dbAuthority = Uri.Authority(
+        host = Uri.Host(config.dbHost),
+        port = config.dbPort.toInt
+    )
+
+    val proxyPort = 15975
+    val proxyActor = actorSystem.actorOf(Props(new RESTProxy("0.0.0.0", proxyPort)(dbAuthority, config.dbProtocol == "https")))
+
+    // the (fake) DB name doesn't matter as this test only execises the "/" endpoint ("instance info")
+    val client = CouchDbRestClient.make("http", "localhost", 15975, config.dbUsername, config.dbPassword, "fakeDB")
+
+    def checkResponse(response: Either[StatusCode,JsObject]) : Unit = response match {
+        case Right(obj) =>
+            assert(obj.fields.contains("couchdb"), "response object doesn't contain 'couchdb'")
+
+        case Left(code) =>
+            assert(false, s"unsuccessful response (code ${code.intValue})")
+    }
+
+    behavior of "CouchDbRestClient"
+
+    it should "successfully access the DB instance info" in {
+        assume(config.dbProvider == "Cloudant" || config.dbProvider == "CouchDB")
+        val f = client.instanceInfo()
+        whenReady(f) { e => checkResponse(e) }
+    }
+
+    it should "successfully access the DB despite transient connection failures" in {
+        // sprays the client with requests, makes sure they are all answered
+        // despite temporary connection failure.
+        val numRequests = 30
+        val timeSpan = 5.seconds
+        val delta = timeSpan / numRequests
+
+        val promises = Vector.fill(numRequests)(Promise[Either[StatusCode,JsObject]])
+
+        for(i <- 0 until numRequests) {
+            actorSystem.scheduler.scheduleOnce(delta * (i+1)) {
+                client.instanceInfo().andThen({ case r => promises(i).tryComplete(r) })
+            }
+        }
+
+        // Mayhem! Havoc!
+        actorSystem.scheduler.scheduleOnce(2.5.seconds, proxyActor, RESTProxy.UnbindFor(1.second))
+
+        // What a type!
+        val futures: Vector[Future[Try[Either[StatusCode,JsObject]]]] =
+            promises.map(_.future.map(e => Success(e)).recover { case t: Throwable => Failure(t) })
+
+        whenReady(Future.sequence(futures), Timeout(timeSpan * 2)) { results =>
+            // We check that the first result was OK
+            // (i.e. the service worked before the disruption)
+            results.head.toOption shouldBe defined
+            checkResponse(results.head.get)
+
+            // We check that the last result was OK
+            // (i.e. the service worked again after the disruption)
+            results.last.toOption shouldBe defined
+            checkResponse(results.last.get)
+
+            // We check that there was at least one error
+            // (i.e. we did manage to unbind for a while)
+            results.find(_.isFailure) shouldBe defined
+        }
```

Comment:
`Future.sequence` will time out if any are stuck, and `whenReady` will then
report an error.

On Jul 24, 2016 10:33, "rodric rabbah" <notifications@github.com> wrote:

> In tests/src/whisk/core/database/test/CouchDbRestClientTests.scala
> <https://github.com/openwhisk/openwhisk/pull/900#discussion_r71989613>:
>
> > +
> > +        whenReady(Future.sequence(futures), Timeout(timeSpan * 2)) { results =>
> > +            // We check that the first result was OK
> > +            // (i.e. the service worked before the disruption)
> > +            results.head.toOption shouldBe defined
> > +            checkResponse(results.head.get)
> > +
> > +            // We check that the last result was OK
> > +            // (i.e. the service worked again after the disruption)
> > +            results.last.toOption shouldBe defined
> > +            checkResponse(results.last.get)
> > +
> > +            // We check that there was at least one error
> > +            // (i.e. we did manage to unbind for a while)
> > +            results.find(_.isFailure) shouldBe defined
> > +        }
>
> should the test check that all requests are either success or failure (ie
> none are stuck/incomplete)?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwhisk/openwhisk/pull/900/files/81306de200f919de413b73f2ad4c290620607cb4#r71989613>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO_KzTh6tV4gHaHH0y9gnrPwfDML59yks5qY3fHgaJpZM4JNrIV>
> .
>